crates/language_models/src/provider/mistral.rs

```diff
@@ -131,18 +142,58 @@ impl State {
             Ok(())
         })
     }
+
+    fn fetch_models(&mut self, cx: &mut Context<Self>) -> Task<Result<()>> {
+        let settings = &AllLanguageModelSettings::get_global(cx).mistral;
+        let http_client = self.http_client.clone();
+        let api_url = settings.api_url.clone();
+        let api_key_result = self
+            .api_key
+            .clone()
+            .ok_or_else(|| anyhow!("API key not set. Please authenticate before fetching models."));
+
+        // As a proxy for the server being "authenticated", we'll check if its up by fetching the models
+        cx.spawn(async move |this, cx| {
+            let api_key = api_key_result?;
+            // This now uses the optimized get_models function that performs parallel API requests
+            let models = mistral::fetch_models(http_client.as_ref(), &api_url, &api_key).await?;
+
+            this.update(cx, |this, cx| {
+                this.models = Some(models);
+                cx.notify();
+            })
+        })
+    }
+
+    fn restart_fetch_models_task(&mut self, cx: &mut Context<Self>) {
+        let task = self.fetch_models(cx);
+        self.fetch_models_task.replace(task);
+    }
 }
 
 impl MistralLanguageModelProvider {
     pub fn new(http_client: Arc<dyn HttpClient>, cx: &mut App) -> Self {
         let state = cx.new(|cx| State {
             api_key: None,
             api_key_from_env: false,
-            _subscription: cx.observe_global::<SettingsStore>(|_this: &mut State, cx| {
+            http_client: http_client.clone(),
+            models: None,
+            fetch_models_task: None,
+            _subscription: cx.observe_global::<SettingsStore>(|this: &mut State, cx| {
+                // Restart the fetch_models task when settings change
+                this.restart_fetch_models_task(cx);
                 cx.notify();
             }),
         });
 
+        if let Ok(env_api_key) = std::env::var(MISTRAL_API_KEY_VAR) {
```

Comment:
Why do we need this?

Reply:
My bad, that environment variable check was a leftover from debugging the automatic model loading feature. It's no longer needed, we can remove it.

Reply:
Removed it. Thanks for pointing it out.